### DISCUSSION TITLE
React Native 相关依赖库由 unimodules 升级至 expo

### DISCUSSION THREAD

**Discussion by zhiqingchen** [NervJS/taro#11565] [RFC]
- 提案时间: (2022-04-02)
- 影响版本: (3.x)

## 概述

Taro React Native 相关依赖库由 unimodules 升级至 expo。

0.67 版本作为过渡，提供 unimodules 和 expo 双支持。

0.68 版本以后，仅保留 expo 支持。

## 动机

Expo 是 React Native 生态中的重要角色，提供了比较多的优秀模块，在 Taro 中有较广泛的使用，如 `expo-av` `expo-camera` `expo-barcode-scanner` 等，将来我们还会持续接入新的模块。

从 Expo SDK 43 开始，react-native-unimodules 包已被弃用，模块系统和自动链接的实现现在改用 expo 包。其架构变更原因可参考该文章 [What’s new in Expo modules infrastructure](https://blog.expo.dev/whats-new-in-expo-modules-infrastructure-7a7cdda81ebc)。

React Native 的生态，仍然处于高速迭代的状态。Taro 的策略是尽可能地与其保持一致节奏，以便开发者更快地获得新特性。

## 使用案例

将 Taro 升级至最新版，将原生代码根据相关pr进行变更。

### 规范原生依赖

> 计划 3.4.5

- [x] 原 unimodules 依赖写法不规范，规范为 peerDependencies。
- [x] "@unimodules/core", "@unimodules/react-native-adapter", "unimodules-permissions-interface" 为不必要依赖，删除或移动至 devDependencies。
- [x] "react-native-unimodules" 移至 peerDependencies。

### Taro 主仓库升级至 expo

> 计划 3.5.0

- [x] 将 unimodules 替换为 expo@45。jest-expo 替换为 jest-expo@45。
- [x] 将 `expo-av`，`expo-barcode-scanner`，`expo-sensors` 等所有依赖升级至最新版。
- [x] 解决升级后问题。
- [x] 保证向下兼容，开发者 APP 如只升级 Taro 至新版本，通过 resolution 将相关依赖锁定，仍可正常运行。
- [ ] 版本发布添加 breaking change 声明。

### Taro Native Shell 升级至 expo

> 计划 3.5.0

- [x] 0.67.0 folk 为 0.67.0-unimodules。0.67.0-expo 分支为 expo 版本使用。
- [x] gradle 升级至 7.2。
- [x] 移除 react-native-unimodules 相关原生代码。
- [x] 参考 https://docs.expo.dev/bare/installing-expo-modules/ 加入 Expo。

### Taro Project Template 升级至 expo

> 计划 3.5.0

- [x] template 项目新增 3.5 分支。为 expo 使用。
- [x] 更新代码同 Taro Native Shell 升级至 expo。

### Taro Playground 升级至 expo

> 计划 3.5.0

- [x] 中版本升级 v1.5.0。
- [x] taro 使用 3.5.0。
- [x] 更新代码同 Taro Native Shell 升级至 expo。

## 缺陷

- 老项目升级有一定的迁移成本，升级过程中会出现较多原生相关问题，影响用户升级。
- js 部分需保持两个版本的壳功能兼容。
- **IPHONEOS_DEPLOYMENT_TARGET 由 11.0 改为 12.0。** 需要兼容 11.0 的开发者，需要保留在低版本。

## 替代选择

N/A

## 适配策略

- Taro 主工程支持 expo 版本。开发者如需兼容 unimodules 版本，需通过 resolution 锁定相关依赖版本。

**Comment by zhiqingchen** [comment]
### Taro Native Shell 升级至 expo

https://github.com/NervJS/taro-native-shell/tree/0.67.0-expo

### 0.68 支持

https://github.com/NervJS/taro-native-shell/tree/0.68.0

**Comment by zhiqingchen** [comment]
### Taro Project Template 升级至 expo

3.5 以下会使用

https://github.com/NervJS/taro-project-templates/tree/v3.4/react-native

3.5 会使用

https://github.com/NervJS/taro-project-templates/tree/v3.5/react-native

**Reply by zhiqingchen** [reply]
需要先全局安装 3.5.0 的cli
```
yarn global add @tarojs/cli@alpha
```

**Comment by zhiqingchen** [comment]
如使用的 taro 版本与 壳工程不对应，则需要锁定版本。

### unimodules
```json
{
  "resolutions": {
    "expo-av": "~9.2.3",
    "expo-barcode-scanner": "~10.2.2",
    "expo-brightness": "~9.2.2",
    "expo-camera": "~11.2.2",
    "expo-file-system": "~11.1.3",
    "expo-image-picker": "~10.2.3",
    "expo-keep-awake": "~9.2.0",
    "expo-location": "~12.1.3",
    "expo-permissions": "~12.1.1",
    "expo-sensors": "~10.2.2"
  }
}
```

### expo （RN 67）

```json
{
  "resolutions": {
    "expo-av": "~11.2.3",
    "expo-barcode-scanner": "~11.3.0",
    "expo-brightness": "~10.2.0",
    "expo-camera": "~12.2.0",
    "expo-file-system": "~14.0.0",
    "expo-image-picker": "~13.1.1",
    "expo-keep-awake": "~10.1.1",
    "expo-location": "~14.0.2",
    "expo-sensors": "~11.2.0",
  }
}
```

### expo （RN 68）

```json
{
  "resolutions": {
    "expo-av": "~11.2.3",
    "expo-barcode-scanner": "~11.3.0",
    "expo-brightness": "~10.2.0",
    "expo-camera": "~12.2.0",
    "expo-file-system": "~14.0.0",
    "expo-image-picker": "~13.1.1",
    "expo-keep-awake": "~10.1.1",
    "expo-location": "~14.0.2",
    "expo-sensors": "~11.3.0",
  }
}
```

**Comment by zhiqingchen** [comment]
### Taro Playground 升级至 expo

分支 https://github.com/wuba/taro-playground/tree/releases/v1.5.0

**Reply by zhiqingchen** [reply]
## bugs

0. 安卓无法加载远程 bundle。使用 change bundle location 切换。
1. iOS video & camera 失效 https://github.com/expo/expo/issues/17395

**Reply by iambool** [reply]
> ### Taro Playground 升级至 expo
> 分支 https://github.com/wuba/taro-playground/tree/releases/v1.5.0

ScrollView组件报错
<img width="191" alt="image" src="https://user-images.githubusercontent.com/30134533/167421844-3a664029-3d09-4d31-9993-38c4575dd89d.png">

安卓Camera组件页面图片样式定义包含 display: block; 引发报错
<img width="222" alt="image" src="https://user-images.githubusercontent.com/30134533/167421945-a6d02743-ff1d-4fa3-9abc-1a0a40e01495.png">

**Reply by zhiqingchen** [reply]
v1.5.0-beta.8 已修复
1. ScrollView组件报错
2. 安卓Camera组件页面图片样式定义包含 display: block; 引发报错
3. iOS video & camera 失效

**Comment by zhiqingchen** [comment]
### Taro 主仓库升级至 expo

3.5 分支升级

https://github.com/NervJS/taro/tree/feat/3.5
